katsdpsigproc/tune.py

```diff
@@ -139,49 +146,31 @@ def _query(
         Keys and values for the query
     """
     try:
-        query, query_args = _assemble_query(tablename, keys)
-        cursor = conn.cursor()
-        cursor.execute(query, query_args)
-        row = cursor.fetchone()
-        # if a match is not found, find the nearest match by first ignoring the device version
-        if row is None:
-            _logger.debug(
-                "Retrying query '%s' by ignoring device_version", query, exc_info=True
-            )
-            try:
-                del keys["device_version"]
-            except KeyError:
-                pass
-            query, query_args = _assemble_query(tablename, keys)
-            cursor = conn.cursor()
-            cursor.execute(query, query_args)
-            row = cursor.fetchone()
-        # if a match is still not found, next try ignoring the device platform
+        query, query_args, row = _assemble_query(conn, tablename, keys)
         if row is None:
-            _logger.debug(
-                "Retrying query '%s' by ignoring device_platform", query, exc_info=True
-            )
-            try:
-                del keys["device_platform"]
-            except KeyError:
-                pass
-            query, query_args = _assemble_query(tablename, keys)
-            cursor = conn.cursor()
-            cursor.execute(query, query_args)
-            row = cursor.fetchone()
-        # if a match is still not found, next try ignoring the device name
-        if row is None:
-            _logger.debug(
-                "Retrying query '%s' by ignoring device_name", query, exc_info=True
-            )
-            try:
-                del keys["device_name"]
-            except KeyError:
-                pass
-            query, query_args = _assemble_query(tablename, keys)
-            cursor = conn.cursor()
-            cursor.execute(query, query_args)
-            row = cursor.fetchone()
+            if KATSDPSIGPROC_TUNE_FALLBACK == "fail":
+                raise RuntimeError(
+                    "Did not find record and autotune fallback behaviour set to 'fail'."
+                )
+            elif KATSDPSIGPROC_TUNE_FALLBACK == "nearest":
+                """If it exists, find the nearest autotune match by first ignoring device version,
+                    then device platform, then device name."""
+                tune_keys = dict(keys)
+                for key in ["device_version", "device_platform", "device_name"]:
+                    _logger.debug("Retrying query '%s' by ignoring %s", query, key)
+                    try:
+                        del tune_keys[key]
+                    except KeyError:
+                        _logger.debug("Key: %s not found in %s", key, tablename, exc_info=True)
```

Comment:
Is this code reachable? `_query` is always called passing it `_db_keys`, and `_db_keys` adds those keys unconditionally.